### PR TITLE
Rework datum infrastructure

### PR DESCRIPTION
**Pull request by c42f**
Originally this PR was for GDA94 and ITRF datum transformations, but it got completely out of control! I have

* Written a lot of documentation
* Reworked the datum types, adding a new abstract `Datum` type.
* Added a datum shift function, `datum_shift_ECEF`, which depends on the datum type
* Added the Dawson et al. GDA94 <--> ITRF datum transformations

---
Original comment -

Still a work in progress:

* The tests suggest something's still wrong here, needs further testing.  Unfortunately I don't know of an easy source of good test data for this Australia wide.  Perhaps we can get some current station coordinates from ftp://ftp.ga.gov.au/geodesy-outgoing/gnss/solutions/final/weekly/ and compare to nominal GDA94 station coordinates.  May require some data munging.
* Depends on CoordinateTransformations affine transformation branch

@andyferris

### CODE REVIEW

**Comment by c42f** [comment]
This is pretty much good to go now - just needs a little more docs and for the affine transformations branch to be finished.

**Comment by andyferris** [comment]
See https://github.com/FugroRoames/CoordinateTransformations.jl/pull/9

**Comment by andyferris** [comment]
LVGTM

I'm noticing that so many transformations are actually constructors for other transformations. It's an interesting trend.

**Comment by c42f** [comment]
Oh, I think I've got a good idea to make the API here a lot nicer!

```julia

immutable ITRF{realization_year, EpochT}
    epoch::EpochT
end

immutable GDA94; end

# Create a transform from GDA94 to ITRF2008 at epoch 2016-1-1
trans = geocentric_datum_transform(ITRF{2008}(Date(2016)), GDA94())

# Create a transform from ITRF2008 to GDA94 which requires a time in the input coordinate (ENUt?)
trans = geocentric_datum_transform(GDA94(), ITRF{2008}(nothing))
```

Some nice things:
* Datums are types.  I think this is "the right thing"
* A consistent way to deal with dynamic vs static datums
* We dodge the coordinate system issues by naming it in the function name ("geocentric", or maybe use ECEF)

**Comment by andyferris** [comment]
Yes, looks really good! I like all of those points about it.

So `geodetic_datum_transform()` could return any kind of transformation object? Like an `AffineTransformation` or something more complex when necessary?

**Comment by c42f** [comment]
> So geodetic_datum_transform() could return any kind of transformation object?

Right exactly.  And because the input Datums are types, the output will be type-inferrable.  Here's another cool side effect: Suppose you wanted input and output in LLA instead.  Easy!

```julia
function make_datum_transform_LLA(dest_datum, source_datum)
    LLAfromECEF(ellipsoid(dest_datum)) ∘ make_datum_transform_ECEF(dest_datum, source_datum) ∘ ECEFfromLLA(ellipsoid(source_datum))
end
```

**Comment by c42f** [comment]
Oh, I've been playing with naming for the main function there.  Some more candidate names

```julia
geocentric_datum_transform
datum_transform_ECEF
make_datum_transform_ECEF
datum_shift_ECEF #< Maybe nice?
make_datum_shift_ECEF
```

**Comment by andyferris** [comment]
Yes, compositions work nicely. :)

Another way might be to let the transformation convert to ECEF it it needs to. After all, it knows what datum it is coming from and going to so there is no point in having the user defining the ellipsoid again just to come from LLA/etc.

**Comment by andyferris** [comment]
Not sure which name works best. I think it's wise to make sure it seems like a function that *creates* a transformation, not *applies* a transformation.

**Comment by andyferris** [comment]
What about `datum_shift()` ?

**Comment by andyferris** [comment]
What are the frames that datum shifts might occur in?

**Comment by c42f** [comment]
The best known version of a datum shift may be defined in any coordinate system.  Eg for OSGB36 this is some projected coordinate system, I'm not sure about the details yet.

**Comment by c42f** [comment]
Some discussion of the ITRF vs ITRS confusion - Kovalevsky J, Mueller II, Kolaczek B (eds) (1989) "Reference frames in astronomy and geophysics" [google books link](https://books.google.com.au/books?hl=en&lr=&id=P7nrCAAAQBAJ&oi=fnd&pg=PA2&dq=Kovalevsky+J,+Mueller+II,+Kolaczek+B+(eds)+(1989)+Reference+frames+in+astronomy+and+geophysics&ots=UNbkJof3SE&sig=AvVQWs1BLQEi73fcL43nLWREXXw&redir_esc=y#v=onepage&q&f=false)

Following is probably worth a read too: "ITRF2008: an improved solution of the international terrestrial reference frame" <http://dx.doi.org/10.1007/s00190-011-0444-4>

File: README.md [hunk 1]
```diff
@@ -97,6 +98,139 @@ distance(x_lla, y_lla)                   # 401.54 meters
...
+authority should also be included where possible.
+
+### Datums
+
+In spatial measurement and positioning, a **datum** is a set of reference
+objects and assigned coordinates *relative to which* other objects may be
+positioned.  For example, in traditional surveying a datum might comprise a
+pair of pegs in the ground, separated by a carefully measured distance.  When
+surveying the position of an unknown but nearby point, the angle back to the
+original datum objects can be measured using a theodolite.  After this, the
```
Comment (awbsmith) [inline README.md]: Did you mean troglodyte?

Comment (c42f) [inline README.md]: lol :-)

@PaulBellette suggests trilobite

**Comment by c42f** [comment]
More resources

Reference frames in practice
http://www.fig.net/resources/publications/figpub/pub64/Figpub64.pdf

ITRF2008 -> earlier ITRF transformation parameters
http://itrf.ensg.ign.fr/doc_ITRF/Transfo-ITRF2008_ITRFs.txt
The IGN also provides a datum shift service which could be used for test data.

Article about ITRF2014
http://gpsworld.com/nasa-helps-maintain-international-terrestrial-reference-frame-with-gnss/

File: src/datum_transformations.jl [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+"""
+    datum_shift_ECEF(dest_datum, source_datum)
+
+Return a transformation object which transforms ECEF points from `source_datum`
+to `dest_datum`.  The function should attempt to supply the current publicly
+accepted best estimate.
+
+Note that the best known version of a datum transformation will inherently
+improve with time (see below), so we *cannot simultaneously* guarentee that:
```
Comment (PaulBellette) [inline src/datum_transformations.jl]: typo on guarantee

Comment (c42f) [inline src/datum_transformations.jl]: you must have been reading an older version, I already fixed this yesterday ;-)

Thanks though!  (the spellchecker also let me know that I can't spell Britain)

File: src/datum_transformations.jl [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    params, rates = table[ITRF_year]
+
+    # Fractional years since reference epoch.  Seems a bit unnecessarily cumbersome!
+    millisecs_per_year = 365.25 * Dates.value(Dates.Millisecond(Dates.Day(1)))
+    dt = Dates.value(Dates.Millisecond(convert(DateTime, epoch) - reference_epoch)) / millisecs_per_year
+
+    # Convert units to meters and radians
+    mas2rad = deg2rad(1e-3/(60*60))
+    unitconv = [1e-3, 1e-3, 1e-3, 1e-9, mas2rad, mas2rad, mas2rad]
+    Tx,Ty,Tz, Sc, Rx,Ry,Rz = unitconv .* (params .+ rates*dt)
```
Comment (awbsmith) [inline src/datum_transformations.jl]: Is it worth worrying about the allocations here?

Comment (c42f) [inline src/datum_transformations.jl]: Nope, not worth worrying about for my usage - I'm going to use this function once for a given date, and that'll return an `AffineTransformation`.  Then apply this transformation to many points.

For a truly dynamic datum transform we still need to figure out how to work with time attached to points, in which case we need to rework this a little.

**Comment by c42f** [comment]
Ok, the documentation turned out to be much more effort than I expected, but I learned a lot about geodesy doing it!  For now, I think I'm done.

Tests are still appearing to fail due to CoordinateTransformations (need to tag that), but work locally on 0.4.

The only thing left to do here might be figuring out what to do about passing instances vs types of Datums.  It seems like instances should be the way to go, especially since some datums like ITRF seem to require that for good design, but I'm somewhat unhappy about the naming options - eg, GDA94 looks very odd to me in lowercase as gda94.

**Comment by c42f** [comment]
@andyferris  I think I sorted out the Ellipsoid pretty printing problem:

```julia
julia> LLAfromECEF(GDA94())
LLAfromECEF(Ellipsoid(grs80))

julia> LLAfromECEF(Ellipsoid(grs80))
LLAfromECEF(Ellipsoid(grs80))
```

**Comment by c42f** [comment]
I think this is in pretty good shape now after a few final tweaks, I think everyone is on the same page with it, so I'm going to merge.

Tests still failing due to CoordinateTransformations not being tagged, but work locally.

**Comment by andyferris** [comment]
Great!

**Comment by andyferris** [comment]
So what should we do about CoordinateTransformations? I'm thinking something along the lines of what I'm planning for Rotations: tag a 0.4-compatible version with AffineTransformation and make a seperate tag for 0.5. This will continue to propagate here, since we have some `FixedVectorNoTuple`s... (I just spotted a mistake in the `REQUIRE`)

**Comment by andyferris** [comment]
Regarding my last post, would you like to tag this?

**Comment by c42f** [comment]
I'm keen to to tag CoordinateTransformations soon, but it might be worth considering settling on a couple of the API-breaking things before we do so.

**Pull request merged**